Hazelcast.Net/Hazelcast.Client.Spi/ClientListenerService.cs

```diff
@@ -14,113 +14,394 @@
+            Debug.Assert(!Thread.CurrentThread.Name.Contains("eventRegistration"));
+
+            ConcurrentDictionary<ClientConnection, EventRegistration> registrationMap;
+            if (_registrations.TryGetValue(registrationKey, out registrationMap) &&
+                registrationMap.ContainsKey(connection))
+            {
+                return;
+            }
+            var future = ((ClientInvocationService) _client.GetInvocationService())
+                .InvokeListenerOnConnection(registrationKey.RegistrationRequest, registrationKey.EventHandler,
+                    connection);
+
+            IClientMessage clientMessage;
+            try
+            {
+                clientMessage = ThreadUtil.GetResult(future);
             }
             catch (Exception e)
             {
                 throw ExceptionUtil.Rethrow(e);
             }
+
+            var serverRegistrationId = registrationKey.ResponseDecoder(clientMessage);
+            var correlationId = registrationKey.RegistrationRequest.GetCorrelationId();
+            var registration = new EventRegistration(serverRegistrationId, correlationId, connection);
+
+            Debug.Assert(registrationMap != null, "registrationMap should be created!");
+            registrationMap.TryAdd(connection, registration);
+        }
+
+        public bool AddEventHandler(long correlationId, DistributedEventHandler eventHandler)
+        {
+            return _eventHandlers.TryAdd(correlationId, eventHandler);
+        }
+
+        public bool RemoveEventHandler(long correlationId)
+        {
+            DistributedEventHandler removed;
+            return _eventHandlers.TryRemove(correlationId, out removed);
+        }
+
+        private void TrySyncConnectToAllConnections()
+        {
+            if (!IsSmart) return;
+
+            // Define the cancellation token.
+            var source = new CancellationTokenSource();
+            var token = source.Token;
+
+            var clientClusterService = _client.GetClientClusterService();
+            var memberList = clientClusterService.GetMemberList();
+            var tasks = memberList.Select(member => Task.Factory.StartNew(() =>
+            {
+                do
+                {
+                    try
+                    {
+                        _connectionManager.GetOrConnect(member.GetAddress());
+                        return;
+                    }
+                    catch (Exception)
+                    {
+                        // ignored as this exception is retryed
+                    }
+                } while (_client.GetLifecycleService().IsRunning());
+            }, token)).ToArray();
+
+            var timeoutMillis = ((ClientInvocationService) _client.GetInvocationService()).InvocationTimeoutMillis;
+            if (!Task.WaitAll(tasks, timeoutMillis))
+            {
+                throw new TimeoutException("Registering listeners is timed out.");
```

Comment:
If task timeouts, then is there a way to force kill the task?

Reply:
cancellation added